src/v2/kuadrant/pipeline/tasks/ratelimit.rs

```diff
@@ -1,36 +1,766 @@
+        match self.expression.eval(ctx) {
+            Ok(AttributeState::Available(value)) => {
+                let value_str = match value {
+                    Value::Int(n) => format!("{n}"),
+                    Value::UInt(n) => format!("{n}"),
+                    Value::Float(n) => format!("{n}"),
+                    Value::String(s) => s.to_string(),
+                    Value::Bool(b) => format!("{b}"),
+                    Value::Null => "null".to_owned(),
+                    _ => {
+                        return Err(EvaluationError::new(
+                            self.expression,
+                            "Only scalar values can be sent as data".to_string(),
+                        ));
+                    }
+                };
+
+                Ok(Entry {
+                    key: self.key.clone(),
+                    value: value_str,
+                })
+            }
+            Ok(AttributeState::Pending) => Err(EvaluationError::new(
+                self.expression,
+                format!("Expression '{}' evaluation is pending", self.key),
+            )),
+            Err(cel_err) => Err(EvaluationError::new(
+                self.expression,
+                format!("CEL evaluation error for '{}': {}", self.key, cel_err),
+            )),
+        }
+    }
+}
+
+const KNOWN_ATTRIBUTES: [&str; 2] = ["ratelimit.domain", "ratelimit.hits_addend"];
+
+/// A task that performs rate limiting by sending descriptors to a rate limit service
+pub struct RateLimitTask {
+    task_id: String,
+    dependencies: Vec<String>,
+    pauses_filter: bool,
+
+    // Rate limit configuration
     scope: String,
-    predicate: Predicate,
-    service: Rc<dyn Service<Response = bool>>,
-    allow_task: Option<Box<dyn Task>>,
-    deny_task: Box<dyn Task>,
+    service: Rc<RateLimitService>,
+
+    // Conditional data for building descriptors
+    conditional_data_sets: Vec<ConditionalData>,
+    predicates: Vec<Predicate>,
+
+    // Default hits addend
+    default_hits_addend: u32,
+}
+
+#[allow(dead_code)]
+impl RateLimitTask {
+    pub fn new(
+        task_id: String,
+        dependencies: Vec<String>,
+        service: Rc<RateLimitService>,
+        scope: String,
+        predicates: Vec<Predicate>,
+        conditional_data_sets: Vec<ConditionalData>,
+    ) -> Self {
+        Self {
+            task_id,
+            dependencies,
+            pauses_filter: true,
+            scope,
+            service,
+            predicates,
+            conditional_data_sets,
+            default_hits_addend: 1,
+        }
+    }
+
+    /// Builds the rate limit descriptors from the context
+    fn build_descriptors(
+        &self,
+        ctx: &ReqRespCtx,
+    ) -> Result<AttributeState<Vec<RateLimitDescriptor>>, EvaluationError> {
+        // TODO: Candidate for a `prepare` task/method
+        if !self.predicates.is_empty() {
+            for predicate in &self.predicates {
+                match predicate.test(ctx)? {
```

Comment:
Any reason why this logic is not using the exposed `predicates.apply()` method but iterates through them all? It's nice to have that encapsulated in one place and can match on the outcome here